pkg/notes/toc.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package notes
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"regexp"
+	"strings"
+)
+
+var headerPattern = regexp.MustCompile("^(?P<indent>#+) ?(?P<title>.+)$")
+
+func GenerateTOC(input string) string {
+	result := &strings.Builder{}
+
+	prevLine := ""
+	lastLen, indent := 0, 0
+	headers := map[string]int{}
+
+	scanner := bufio.NewScanner(strings.NewReader(input))
+	for scanner.Scan() {
+		if headerPattern.Match(scanner.Bytes()) {
+			matches := headerPattern.FindStringSubmatch(scanner.Text())
+
+			// skip code blocks completely
+			if strings.Contains(prevLine, "`") {
+				continue
+			}
```

Comment:
Not sure what that is about, TBH. Wouldn't that skip the 2nd of two successive headings, if the 1st one had a backtick in it's title?
```text
# This has `some backticks`
## And this heading will be skipped now?
```

There is no test case for that either so I am not sure about the use-case / intention for that special case.

Reply:
Actually, wouldn't it break if we'd had a header line with a backtick in it?
- we find a header and set `prevLine` to it's content, coincidentally it contains a backtick
- next time around the if clause matches and we `continue`, `prevLine` still is set to the thing with the backtick
- next time around the if clause matches and we `continue`, `prevLine` still is set to the thing with the backtick
- ...
- eventually we are done scanning, but never really processed much except up until the first header with a backtick

Reply:
The main purpose of this is to skip code blocks (do not generate any headers inside them, because they could contain `#`. :) I think we have to somehow verify if we're inside a code block by the amount of backticks we've seen. :thinking:

Reply:
Tracking the right amount of back ticks should fix the code block issue in a more general way, PTAL :)

Reply:
I don't think this is working correctly. The current implementation would ignore a perfectly valid heading
```
# `markdown` solves all our problems, they said
```

Other issues might arise with documents like
````````````
something

``````
wow ... this actually seems to be valid md

# ````go
# hello := `hello`
# ````
``````

something else
````````````
where the number of seen backticks is even in every line. This example is contrived, however we are dealing with random user input here.

Honest questions:
- How much effort do we want to spend here for correctly rendering the TOC?
- How likely is it that a b0rken TOC blanks out or scrambles the rest of the document?

Reply:
The first example should be covered by the current implementation, I verified that with an additional unit test:
https://github.com/kubernetes/release/blob/5f80175a27448cc129ca0e800f89b1de0687fb98/pkg/notes/toc_test.go#L148-L152

The second example is tricky, right. So what do you prefer? I could remove the overall code block check vs leaving it as it currently is.